### PR TITLE
openshift_checks: enable writing results to files

### PR DESCRIPTION
**Pull request by sosiouxme**
An iteration on how to record check results in a directory structure readable by machines and humans.

Some refactoring of checks and the action plugin to enable writing files
locally about the check operation and results, if the user wants them.
This is aimed at enabling persistent and machine-readable results from
recurring runs of health checks.

Now, rather than trying to build a result hash to return from running
each check, checks can just register what they need to as they're going
along, and the action plugin processes state when the check is done.
Checks can register failures, notes about what they saw, and arbitrary
files to be saved into a directory structure where the user specifies.
If no directory is specified, no files are written.

At this time checks can still return a result hash, but that will likely
be refactored away in the next iteration.

Multiple failures can be registered without halting check execution.
Throwing an exception or returning a hash with "failed" is registered as
a failure.

execute_module now does a little more with the results. Results are
automatically included in notes and written individually as files.
"changed" results are propagated. Some json results are decoded.

A few of the checks were enhanced to use these features; all get some of
the features for free.

Action items:

- [x] Provide a way for user to specify an output directory where they want results written
- [x] Enable a check to register multiple failures and not have to assemble them in result
- [x] Enable a check to register "notes" that will be saved to files but not displayed
- [x] Have module invocations recorded individually as well as in notes
- [x] Enable a check to register files (logs, etc.) from remote host that are to be copied to output dir
- [x] Enable a check to register arbitrary file contents that are to be written to output
- [ ] Take advantage of these features where possible in checks

(Last item done somewhat, more should happen as we go along...)

### CODE REVIEW

File: roles/openshift_health_checker/action_plugins/openshift_health_check.py [hunk 1]
```diff
@@ -196,17 +201,75 @@ def run_check(name, check, user_disabled_checks):
         is_active = check.is_active()
     except Exception as exc:
         reason = "Could not determine if check should be run, exception: {}".format(exc)
-        return dict(skipped=True, skipped_reason=reason, exception=traceback.format_exc())
+        return dict(skipped=True, skipped_reason=reason, exception=traceback.print_exc())
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/action_plugins/openshift_health_check.py]: I didn't understand the reason for this change. `format_exc` returns a string, `print_exc` writes to a file -- `sys.stderr` by default. IIUC, using `print_exc` here causes the side effect but makes an entry `exception: None` in the returned `dict`.

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 2]
```diff
@@ -94,6 +100,19 @@ def subclasses(cls):
             for subclass in subclass.subclasses():
                 yield subclass
 
+    def register_failure(self, error):
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Why is this part of the `OpenShiftCheck` class? When can a check have more than one failure?

File: roles/openshift_health_checker/action_plugins/openshift_health_check.py (hunk 1 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/action_plugins/openshift_health_check.py]: that's my mistake, I was fiddling with something, got distracted and left it there. thanks!

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 2 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/openshift_checks/__init__.py]: Frequently. For instance, the disk availability check previously would fail on the first file system it found that was too full; now it will report multiple problem paths. The EFK checks often check multiple things and put together a list of failures, and this is actually a bit messy when they're coming from different methods. Easier if you can just record it where it's found and let it be summarized at the end. Of course you can also raise an exception if the check can't proceed after a failure.

File: roles/openshift_health_checker/action_plugins/openshift_health_check.py [hunk 3]
```diff
@@ -201,12 +215,131 @@ def run_check(name, check, user_disabled_checks):
...
+
+def write_to_output_file(output_dir, filename, data):
+    """If output_dir provided, write data to file. Serialize as JSON if data is not a string."""
+
+    if not output_dir or not prepare_output_dir(output_dir):
+        return
+    filename = os.path.join(output_dir, filename)
+    try:
+        with open(filename, 'w') as outfile:
+            if isinstance(data, basestring):
```
Comment (sosiouxme) [inline roles/openshift_health_checker/action_plugins/openshift_health_check.py]: python3 doesn't like basestring. Should I be doing:

    from past.builtins import basestring 

...or something else?

Comment (rhcarvalho) [reply roles/openshift_health_checker/action_plugins/openshift_health_check.py]: https://github.com/openshift/openshift-ansible/blob/20cf3396e4aae5aef6485a23e6cc3134c7a2d1be/roles/openshift_health_checker/action_plugins/openshift_health_check.py#L10
https://github.com/openshift/openshift-ansible/blob/20cf3396e4aae5aef6485a23e6cc3134c7a2d1be/roles/openshift_health_checker/action_plugins/openshift_health_check.py#L183

Comment (sosiouxme) [reply roles/openshift_health_checker/action_plugins/openshift_health_check.py]: thanks. apparently i'm blind, there being already an example in the file.

**Comment by sosiouxme** [comment]
@rhcarvalho if you have time to review, would be appreciated. I think the features I wanted are pretty much there, can expand the use cases as we see the need.

Saved files from a `health.yml` run may look something like:

```
/tmp/health.CnPp4FFv/
├── ec2-54-172-162-93.compute-1.amazonaws.com
│   ├── curator
│   │   └── ocutil.json
│   ├── curator.notes.json
│   ├── elasticsearch
│   │   ├── get_es_health.json
│   │   ├── get_es_nodes.json
│   │   ├── get_master_names.json
│   │   ├── get_pv_diskspace.json
│   │   └── ocutil.json
│   ├── elasticsearch.notes.json
│   ├── etcd_volume.notes.json
│   ├── fluentd
│   │   ├── ocutil.json
│   │   └── ocutil.json.2
│   ├── fluentd.notes.json
│   ├── kibana
│   │   ├── ocutil.json
│   │   ├── ocutil.json.2
│   │   └── uri.json
│   ├── kibana.failures.json
│   ├── kibana.notes.json
│   ├── logging_index_time
│   │   ├── ocutil.json
│   │   ├── ocutil.json.2
│   │   ├── ocutil.json.3
│   │   ├── query_for_uuid.json
│   │   ├── query_for_uuid.json.2
│   │   ├── query_for_uuid.json.3
│   │   ├── query_for_uuid.json.4
│   │   └── query_for_uuid.json.5
│   ├── logging_index_time.notes.json
│   ├── ovs_version
│   │   └── rpm_version.json
│   ├── ovs_version.notes.json
│   └── result.json
└── ec2-54-88-135-189.compute-1.amazonaws.com
    ├── ovs_version
    │   └── rpm_version.json
    ├── ovs_version.notes.json
    └── result.json
```

**Comment by rhcarvalho** [comment]
> ─ curator.notes.json

what are the `.notes.json` files?─ curator.notes.json

File: roles/openshift_health_checker/test/ovs_version_test.py [hunk 4]
```diff
@@ -50,7 +50,7 @@ def test_ovs_package_version(openshift_release, expected_ovs_version):
         openshift_release=openshift_release,
         openshift_image_tag='v' + openshift_release,
     )
-    return_value = object()
+    return_value = {}
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/ovs_version_test.py]: Why this change?

The original is an idiom to say you get the exact same thing when doing comparison.

**Comment by sosiouxme** [comment]
Maybe "notes" isn't the best name. Is sort of a log of everything that happens in sequence; any `execute_module` calls, any failures (`register_failure`), any other info (`register_note`) that the check wants to add about what it saw in making its determination.

File: roles/openshift_health_checker/test/openshift_check_test.py [hunk 5]
```diff
@@ -106,13 +106,39 @@ def test_get_var_convert(task_vars, keys, convert, expected):
...
+
+def test_register(task_vars):
+    check = dummy_check(task_vars)
+
+    check.register_failure(OpenShiftCheckException("spam"))
+    assert "spam" in str(check.failures[0])
+
+    with pytest.raises(OpenShiftCheckException) as excinfo:
+        check.register_file("spam")
+    assert "not specified" in str(excinfo.value)
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/openshift_check_test.py]: looking at the test, it is not obvious why there is an expected exception, what's missing?

File: roles/openshift_health_checker/test/openshift_check_test.py [hunk 6]
```diff
@@ -106,13 +106,39 @@ def test_get_var_convert(task_vars, keys, convert, expected):
...
+
+    check._execute_module = lambda *args, **_: dict()
+    check.execute_module("eggs", module_args={}, register=False)
+    assert not check.files_to_save
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/openshift_check_test.py]: What do this block mean?

I'm guessing execute a module and have nothing to save? How does it fit with the rest of the test function?

File: roles/openshift_health_checker/test/elasticsearch_test.py [hunk 7]
```diff
@@ -72,7 +72,7 @@ def test_check_elasticsearch():
     assert_error_in_list('NoRunningPods', excinfo.value)
 
     # canned oc responses to match so all the checks pass
-    def exec_oc(cmd, args):
+    def exec_oc(cmd, args, **_):
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/elasticsearch_test.py]: Why is this?

File: roles/openshift_health_checker/test/disk_availability_test.py [hunk 8]
```diff
@@ -183,11 +183,12 @@ def test_fails_with_insufficient_disk_space(name, group_names, configured_min, a
         ansible_mounts=ansible_mounts,
     )
 
-    result = DiskAvailability(fake_execute_module, task_vars).run()
+    check = DiskAvailability(fake_execute_module, task_vars)
+    check.run()
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/disk_availability_test.py]: I read the new docstring saying `run` returns a dict but it will be empty and here we ignore the return value altogether. Maybe its less hassle to just get rid of the return value?

From the testing perspective, pure functions returning values tend to be easier to test than side-effects (e.g., attributes), but I'll not go into that discussion :-)

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 9]
```diff
@@ -94,7 +120,30 @@ def subclasses(cls):
...
+        # It should be an exception; make it one if not
+        if not isinstance(error, OpenShiftCheckException):
+            error = OpenShiftCheckException(str(error))
+        self.failures.append(error)
+        # duplicate it in the notes so it can be seen in the context of any
+        # information that led to the failure
+        self.register_note("failure: " + error.name, str(error))
+
+    def register_note(self, context, msg):
+        """Record a note for the check, intended to be saved as an ordered list in a local notes file."""
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: What is the purpose of a note? I think that explanation would be useful to have in the docstring.

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 10]
```diff
@@ -94,7 +120,30 @@ def subclasses(cls):
             for subclass in subclass.subclasses():
                 yield subclass
 
-    def execute_module(self, module_name=None, module_args=None):
+    def register_failure(self, error):
+        """Record failure in the check, to be merged into result hash and saved to output directory."""
+        # It should be an exception; make it one if not
+        if not isinstance(error, OpenShiftCheckException):
+            error = OpenShiftCheckException(str(error))
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Hmm, why is this done?

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 11]
```diff
@@ -94,7 +120,30 @@ def subclasses(cls):
             for subclass in subclass.subclasses():
                 yield subclass
 
-    def execute_module(self, module_name=None, module_args=None):
+    def register_failure(self, error):
+        """Record failure in the check, to be merged into result hash and saved to output directory."""
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Document that registering a failure has a side-effect of registering a note.

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 12]
```diff
@@ -94,7 +120,30 @@ def subclasses(cls):
...
+
+    def register_note(self, context, msg):
+        """Record a note for the check, intended to be saved as an ordered list in a local notes file."""
+        self.notes.append([context, msg])
+
+    def register_file(self, filename, contents=None, remote_filename=""):
+        """Record a file that a check makes available to be saved individually to output directory.
+        Either file contents should be passed in, or a file to be copied from the remote host
+        should be specified. Contents that are not a string are to be serialized as JSON.
+        """
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Aha, nice idea! :+1:

**Comment by sosiouxme** [comment]
Perhaps `<check>.log.json` instead?

**Comment by rhcarvalho** [comment]
> Perhaps <check>.log.json instead?

Yes, that's what I was thinking.

File: roles/openshift_health_checker/test/ovs_version_test.py (hunk 4 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/ovs_version_test.py]: Because now you don't get the exact thing; `execute_module` doctors the result, and it expects a hash returned.

`assert result is return_value` means you get the same hash back... though its contents aren't the same.

File: roles/openshift_health_checker/test/openshift_check_test.py (hunk 5 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/openshift_check_test.py]: some kind of content for the file :) A comment would help

File: roles/openshift_health_checker/test/openshift_check_test.py (hunk 6 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/openshift_check_test.py]: Normally `execute_module` would register a file to save with its results. This is testing that `register=False` disables that in the case where it would just be clutter.

File: roles/openshift_health_checker/test/elasticsearch_test.py (hunk 7 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/elasticsearch_test.py]: Updated some of the checks to add the `save_as_name` kwarg and don't really have an idiom for `lambda ignore-all-args-and-kwargs`

File: roles/openshift_health_checker/test/disk_availability_test.py (hunk 8 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/disk_availability_test.py]: I would prefer to either have a return value or record state, not both. I think the long-term direction is to record state because keeping track of a single return value is tedious (passing between methods, etc) and leads to lots of duplicate logic sprinkled around.

I didn't completely refactor to get rid of return values, but would like to do that.

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 10 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/openshift_checks/__init__.py]: There was one place in the action plugin where it takes the hash result and turns `failed` + `msg` into a registered failure, and I thought it would be simpler to make that generic, to ensure that what's in `failures` is always an `OpenShiftCheckException`.

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 12 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/openshift_checks/__init__.py]: I should document however that this is transferred via slurp, meaning base64 encoded in memory... not something you want to use to transfer your sosreport tarball or anything else huge.

**Comment by sosiouxme** [comment]
@rhcarvalho Thanks, I addressed that much; I imagine you have more...

**Comment by sosiouxme** [comment]
> @openshift-ci-robot added size/XL and removed size/L labels

You callin' me fat?

**Comment by sosiouxme** [comment]
aos-ci-test

**Comment by sosiouxme** [comment]
aos-ci-test

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 13]
```diff
@@ -82,7 +102,13 @@ def is_active():
 
     @abstractmethod
     def run(self):
-        """Executes a check, normally implemented as a module."""
+        """Executes a check against a host and returns a result hash similar to Ansible modules.
+
+        Actually the directon ahead is to record state in the attributes and
+        not bother building a result hash. Instead, return an empty hash and let
+        the action plugin fill it in. Or raise an OpenShiftCheckException.
+        Returning a hash may become deprecated if it does not prove necessary.
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: I would go for getting rid of the old way if possible to avoid confusion

File: roles/openshift_health_checker/test/action_plugin_test.py [hunk 14]
```diff
@@ -166,24 +195,51 @@ def test_action_plugin_run_check_fail(plugin, task_vars, monkeypatch):
...
     assert changed(result)
     assert not skipped(result)
 
 
+def test_action_plugin_run_check_output_dir(plugin, task_vars, tmpdir, monkeypatch):
+    check_class = fake_check(
+        run_return={},
+        run_notes=[('thing', 'note')],
+        run_files=[
+            FileToSave('save.file', 'contents', None),
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/action_plugin_test.py]: Does this create actual files during test?

File: roles/openshift_health_checker/test/action_plugin_test.py [hunk 15]
```diff
@@ -249,3 +305,38 @@ def test_resolve_checks_failure(names, all_checks, words_in_exception):
...
+
+
+@pytest.mark.parametrize('give_output_dir, result, expect_file', [
+    (False, None, False),
+    (True, dict(content="c3BhbQo=", encoding="base64"), True),
+    (True, dict(content="encoding error", encoding="base64"), False),
+    (True, dict(content="spam", no_encoding=None), True),
+    (True, dict(failed=True, msg="could not slurp"), False),
+])
+def test_copy_remote_file_to_dir(give_output_dir, result, expect_file, tmpdir):
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/action_plugin_test.py]: I suspect we could reuse whatever file transfer mechanism that Ansible has?

File: roles/openshift_health_checker/test/action_plugin_test.py [hunk 16]
```diff
@@ -249,3 +305,38 @@ def test_resolve_checks_failure(names, all_checks, words_in_exception):
...
+    check = fake_check()()
+    check.execute_module = lambda *args, **_: result
+    copy_remote_file_to_dir(check, "remote_file", str(tmpdir) if give_output_dir else "", "local_file")
+    assert expect_file == any(path.basename == "local_file" for path in tmpdir.listdir())
+
+
+def test_write_to_output_exceptions(tmpdir, monkeypatch, capsys):
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/action_plugin_test.py]: What is capsys here?

File: roles/openshift_health_checker/action_plugins/openshift_health_check.py [hunk 17]
```diff
@@ -90,7 +97,12 @@ def load_known_checks(self, tmp, task_vars):
                     "duplicate check name '{}' in: '{}' and '{}'"
                     "".format(name, full_class_name(cls), full_class_name(other_cls))
                 )
-            known_checks[name] = cls(execute_module=self._execute_module, tmp=tmp, task_vars=task_vars)
+            known_checks[name] = cls(
+                execute_module=self._execute_module,
+                tmp=tmp,
+                task_vars=task_vars,
+                want_full_results=want_full_results
```
[the remark this answers is not in the document]
Comment (sosiouxme) [reply roles/openshift_health_checker/action_plugins/openshift_health_check.py]:
```
    Optional init param: want_full_results
    If the check can gather logs, tarballs, etc., do so when True; but no need to spend
    the time if they're not wanted (won't be written to output directory).  
```

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 18]
```diff
@@ -40,26 +42,44 @@ def __getitem__(self, index):
...
+        self.want_full_results = want_full_results
 
+        # state to be recorded for inspection after the check runs:
+        #
         # set to True when the check changes the host, for accurate total "changed" count
         self.changed = False
+        # list of OpenShiftCheckException for check to report (alternative to returning a failed result)
+        self.failures = []
+        # notes to include only when writing results to files - tuples of (description, msg)
+        self.notes = []
```
[the remark this answers is not in the document]
Comment (sosiouxme) [reply roles/openshift_health_checker/openshift_checks/__init__.py]: OK will steal my comment. I guess "log" and "logs" would be a lot clearer.

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 19]
```diff
@@ -40,26 +42,44 @@ def __getitem__(self, index):
...
+        # list of FileToSave
+        self.files_to_save = []
```
[the remark this answers is not in the document]
Comment (sosiouxme) [reply roles/openshift_health_checker/openshift_checks/__init__.py]: It doesn't really need to know about them. It made sense to me to have the results of the check go into the check's instance state. It just seems easier conceptually. But I suppose there's no reason they couldn't go to an external logwriter. I'll think about it...

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 13 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/openshift_checks/__init__.py]: Well it's a lot of code and tests to refactor. But maybe I can get it done quickly enough.

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 14 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/action_plugin_test.py]: yes... into a tmp dir

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 16 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/action_plugin_test.py]: pytest thing, captures output and stderr

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 15 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/action_plugin_test.py]: We do... it uses the slurp module.

There is a "fetch" module which sounds like it would be useful but it does not put the file where I want and under the covers it is most likely to be calling slurp anyway. There is a 'synchronize' module which uses rsync if present (I don't want to rely on that on localhost). This is not at all as straightforward as I would expect :(

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 14 above)
Comment (rhcarvalho) [reply roles/openshift_health_checker/test/action_plugin_test.py]: What if we used something like a StringIO buffer, to look like a file in terms of interface, but no disk io

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 15 above)
Comment (rhcarvalho) [reply roles/openshift_health_checker/test/action_plugin_test.py]: Aha, haven't use that myself, unfortunate that is not easy to reuse.

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 16 above)
Comment (rhcarvalho) [reply roles/openshift_health_checker/test/action_plugin_test.py]: Wow, I didn't know it :-)

https://docs.pytest.org/en/latest/capture.html?highlight=capsys#accessing-captured-output-from-a-test-function

File: roles/openshift_health_checker/test/action_plugin_test.py [hunk 20]
```diff
@@ -249,3 +305,38 @@ def test_resolve_checks_failure(names, all_checks, words_in_exception):
...
+
+    test = {1: object(), 2: Spam()}
+    test[3] = test
+    write_result_to_output_dir(str(tmpdir), test)
+    assert "Error writing" in test["output_files"]
+
+    output_dir = tmpdir.join("eggs")
+    output_dir.write("spam")  # so now it's not a dir
+    write_to_output_file(str(output_dir), "somefile", "somedata")
+    assert "Could not write" in capsys.readouterr()[1]
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/action_plugin_test.py]: What is supposed to write to stderr in this case?

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 14 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/action_plugin_test.py]: it creates directories as well as files. it would be fairly complicated to mock that out...

is there a reason to avoid using tmpdir?

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 20 above)
Comment (sosiouxme) [reply roles/openshift_health_checker/test/action_plugin_test.py]: https://github.com/sosiouxme/openshift-ansible/blob/88d68d2e9070f2a62bf9b66773034a56a69c90fc/roles/openshift_health_checker/action_plugins/openshift_health_check.py#L304

        display.warning("Could not write output file {}: {}".format(filename, exc))

Since the problem is with writing a file, there's nowhere to notify the user except in the output.

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 14 above)
Comment (rhcarvalho) [reply roles/openshift_health_checker/test/action_plugin_test.py]: It would require re-architecting the solution, so better just go with `tempdir`, I'm realizing now it is also managed by pytest.

File: roles/openshift_health_checker/test/action_plugin_test.py (hunk 20 above)
Comment (rhcarvalho) [reply roles/openshift_health_checker/test/action_plugin_test.py]: Ok, I get it.

**Comment by sosiouxme** [comment]
@rhcarvalho can I get another LGTM on this?

**Comment by sosiouxme** [comment]
Can I rubberstamp myself? :-P
/lgtm

**Comment by sosiouxme** [comment]
@Miciah I had to fix a bug that somehow wasn't noticed before. Forgot to rename notes => logs in a couple places.
Tests pass now, just needs a LGTM to merge...

**Comment by sosiouxme** [comment]
wot... tests were all green before! something up with CI...

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 21]
```diff
@@ -259,7 +352,7 @@ def load_checks(path=None, subpkg=""):
             modules = modules + load_checks(os.path.join(path, name), subpkg + "." + name)
             continue
 
-        if name.endswith(".py") and not name.startswith(".") and name not in LOADER_EXCLUDES:
+        if name.endswith(".py") and name not in LOADER_EXCLUDES:
```
Comment (Miciah) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: @sosiouxme, somehow I missed this during review.  Why remove the hidden-file check?

Comment (sosiouxme) [reply roles/openshift_health_checker/openshift_checks/__init__.py]: No good reason I can think of; probably a mistake during fixing a merge conflict
